### PR TITLE
Module documentation controller

### PR DESCRIPTION
**Pull request by yalosev**
## Description
Use controller for rendering modules documentation

## Why do we need it, and what problem does it solve?
It's much more convenient to use separate CRD for rending a documentation. 

## What is the expected result?
```
# kubectget moduledocumentations.deckhouse.io
NAME              RESULT
deckhouse-admin   Rendered
echo              Rendered
mcplay            Rendered
parca             Rendered
```

ModuleRelease
```yaml
apiVersion: deckhouse.io/v1alpha1
kind: ModuleDocumentation
metadata:
  creationTimestamp: "2024-05-08T22:57:00Z"
  generation: 1
  labels:
    module: deckhouse-admin
    source: deckhouse-prod
  name: deckhouse-admin
  ownerReferences:
  - apiVersion: deckhouse.io/v1alpha1
    controller: true
    kind: ModuleRelease
    name: deckhouse-admin-v1.23.1
    uid: 4fbd7331-7ed2-4a47-9153-23607c4178e4
  resourceVersion: "151066656"
  uid: a9146eff-18a1-4885-be4a-fc6e67effbee
spec:
  checksum: 6aac80cafa189028e3d19574be0b436a
  path: /deckhouse-admin/v1.23.1
  version: v1.23.1
status:
  conditions:
  - address: http://10-111-2-5.d8-system.pod.cluster.local:8081
    checksum: 6aac80cafa189028e3d19574be0b436a
    lastTransitionTime: "2024-05-08T23:06:01Z"
    message: ""
    type: Rendered
    version: v1.23.1
  result: Rendered
```

ModulePullOverride
```yaml
apiVersion: deckhouse.io/v1alpha1
kind: ModuleDocumentation
metadata:
  creationTimestamp: "2024-05-08T22:58:45Z"
  generation: 3
  labels:
    module: echo
    source: losev-test
  name: echo
  ownerReferences:
  - apiVersion: deckhouse.io/v1alpha1
    controller: true
    kind: ModulePullOverride
    name: echo
    uid: 8e8fe026-3b62-4539-81e9-a5592f35c277
  resourceVersion: "151066664"
  uid: d0c86a90-fe01-4919-a888-76bc5403247d
spec:
  checksum: sha256:46316eaf6a4550190a1df7e5f7b1d68c5b3fe5e7fceae0e5ef50876d6f6dd737
  path: /echo/dev
  version: v0.4.50-patched
status:
  conditions:
  - address: http://10-111-2-5.d8-system.pod.cluster.local:8081
    checksum: sha256:46316eaf6a4550190a1df7e5f7b1d68c5b3fe5e7fceae0e5ef50876d6f6dd737
    lastTransitionTime: "2024-05-08T23:06:01Z"
    message: ""
    type: Rendered
    version: v0.4.50-patched
  result: Rendered
```

Scale documentation deployment up to 2 replicas gives:
```yaml
apiVersion: deckhouse.io/v1alpha1
kind: ModuleDocumentation
metadata:
  creationTimestamp: "2024-05-08T22:58:45Z"
  generation: 3
  labels:
    module: echo
    source: losev-test
  name: echo
  ownerReferences:
  - apiVersion: deckhouse.io/v1alpha1
    controller: true
    kind: ModulePullOverride
    name: echo
    uid: 8e8fe026-3b62-4539-81e9-a5592f35c277
  resourceVersion: "152743041"
  uid: d0c86a90-fe01-4919-a888-76bc5403247d
spec:
  checksum: sha256:46316eaf6a4550190a1df7e5f7b1d68c5b3fe5e7fceae0e5ef50876d6f6dd737
  path: /echo/dev
  version: v0.4.50-patched
status:
  conditions:
  - address: http://10-111-2-5.d8-system.pod.cluster.local:8081
    checksum: sha256:46316eaf6a4550190a1df7e5f7b1d68c5b3fe5e7fceae0e5ef50876d6f6dd737
    lastTransitionTime: "2024-05-08T23:06:01Z"
    message: ""
    type: Rendered
    version: v0.4.50-patched
  - address: http://10-111-2-236.d8-system.pod.cluster.local:8081
    checksum: sha256:46316eaf6a4550190a1df7e5f7b1d68c5b3fe5e7fceae0e5ef50876d6f6dd737
    lastTransitionTime: "2024-05-10T19:04:34Z"
    message: ""
    type: Rendered
    version: v0.4.50-patched
  result: Rendered
```

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: feature
summary: Use CRD for render modules documentation.
impact_level: low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

**Comment by miklezzzz** [comment]
some conflicts have occurred :)

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

**Review by miklezzzz — COMMENTED** [review]
tested in a dev cluster manually and found strange behaviour:
if I delete ./external-modules/<module_name>/<module_version> directory, emulating a FS issue, and then restart a documentation pod, moduledocumentation controller doesn't report any errors on its next attempt to render the documentation.
Meanwhile, there is a new doc lease in the cluster
```
root@dev-master-0:~# kubectl  get leases.coordination.k8s.io -n d8-system 
NAME                        HOLDER                                                                   AGE
module-docs-builder-5vz9f   10-111-2-36.d8-system.pod.cluster.local:8081                             12m
```
but moduledocumentation resource reports stale condition:
```
status:
  conditions:
  - address: http://10-111-2-35.d8-system.pod.cluster.local:8081
    checksum: ca93c777a12a13f5fc509dfdc22c06d2
    lastTransitionTime: "2024-05-11T17:53:24Z"
```

File: deckhouse-controller/crds/doc-ru-module-documentation.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            Определяет конфигурацию для рендеринга документации модуля Deckhouse..
+
+            **Ресурсы ModuleDocumentation создает Deckhouse.**
+          properties:
+            spec:
+              properties:
+                version:
+                  description: Версия модуля.
+                checksum:
+                  description: Чексумма версии модуля.
```
Comment (z9r5) [inline deckhouse-controller/crds/doc-ru-module-documentation.yaml:18]:
```suggestion
                  description: Контрольная сумма версии модуля.
```

**Review by z9r5 — APPROVED** [review]

File: deckhouse-controller/pkg/controller/module-controllers/docbuilder/docs_controller.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+		return ctrl.Result{}, nil
+	}
+
+	pr, pw := io.Pipe()
+
+	mdr.logger.Debugf("Getting the %s module's documentation locally", moduleName)
+	err = mdr.getDocumentationFromModuleDir(md.Spec.Path, pw)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("failed to get %s module documentation from local directory with error: %w", moduleName, err)
+	}
```
[the remark this answers is not in the document]
Comment (yalosev) [reply deckhouse-controller/pkg/controller/module-controllers/docbuilder/docs_controller.go]: Yes. They are downloaded on the other controllers startup. So we don't want to check it here

**Review by nabokihms — APPROVED** [review]

**Pull request merged**